ibtk/src/lagrangian/FEProjector.cpp

```diff
@@ -390,11 +460,23 @@ FEProjector::computeL2Projection(PetscVector<double>& U_vec,
 
     if (close_F) F_vec.close();
     const System& system = d_fe_data->getEquationSystems()->get_system(system_name);
-    const DofMap& dof_map = system.get_dof_map();
-    if (consistent_mass_matrix)
+
+    // We can use the diagonal mass matrix directly if we do not need a
+    // consistent mass matrix *and* there are no constraints.
+    //
+    // TODO: this would also work with Dirichlet boundary constraints but its not
+    // as easy to detect those constraints.
+    if (!consistent_mass_matrix && system.get_dof_map().n_constrained_dofs() == 0)
```

Comment:
👍

Reply:
I wrote something much more complicated at first which is why this took so long!